test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/UrlHelperTest.cs

```diff
@@ -939,6 +939,37 @@ public void Action_RouteValueInvalidation_DoesNotAffectActionAndController_Actio
             Assert.Equal(expected, builder.ToString());
         }
 
+        [Theory]
+        [InlineData(null, null, null, "/", null, "/")]
+        [InlineData(null, null, null, "/", null, "/")]
+        [InlineData(null, null, null, "/Hello", null, "/Hello" )]
+        [InlineData(null, null, null, "/Hello", null, "/Hello")]
+        [InlineData("/", null, null, "", null, "/")]
+        [InlineData("/hello/", null, null, "/world", null, "/hello/world")]
+        [InlineData("/hello/", "https", "myhost", "/world", "fragment-value", "https://myhost/hello/world#fragment-value")]
+        public void GenerateUrl_FastAndSlowPathsReturnsExpected(
+            string appBase,
+            string protocol,
+            string host,
+            string virtualPath,
```

Comment:
How about cases where virtual path doesn't start with `/`? There's a check for it in the product code.